test/erlfmt_markdown_SUITE.erl

```diff
@@ -0,0 +1,180 @@
+    init_per_group/2,
+    end_per_group/2,
+    init_per_testcase/2,
+    end_per_testcase/2
+]).
+
+%% Test cases
+-export([
+    markdown_files/1,
+    markdown_string/1
+]).
+
+suite() ->
+    [{timetrap, {seconds, 10}}].
+
+init_per_suite(Config) ->
+    Config.
+
+end_per_suite(_Config) ->
+    ok.
+
+init_per_group(_GroupName, Config) ->
+    Config.
+
+end_per_group(_GroupName, _Config) ->
+    ok.
+
+init_per_testcase(_TestCase, Config) ->
+    Config.
+
+end_per_testcase(_TestCase, _Config) ->
+    ok.
+
+groups() ->
+    [
+        {markdown, [parallel], [
+            markdown_files,
+            markdown_string
+        ]}
+    ].
+
+all() ->
+    [{group, markdown}].
+
+markdown_files(Config) when is_list(Config) ->
+    {ok, Data} = file:get_cwd(),
+    RepoRootPath = filename:join([Data, "..", "..", "..", ".."]),
+    DocPath = filename:join(RepoRootPath, "doc"),
+    Filenames = lists:append(
+        filelib:wildcard(filename:join(RepoRootPath, "*.md")),
+        filelib:wildcard(filename:join(DocPath, "*.md"))
+    ),
+    ?assertMatch([_One, _Two, _Three | _AtLeast], Filenames),
+    lists:map(
+        fun (Filename) ->
+            {ok, Content} = file:read_file(Filename),
+            check_markdown(binary_to_list(Content))
+        end,
+        Filenames
+    ).
+
+markdown_string(Config) when is_list(Config) ->
+    S =
+        "# Heading\n"
+        "ignore for now\n"
+        "```erlang formatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n"
+        "ignore for now\n"
+        "```erlang unformatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "The previous unformatted version, should match this formatted version\n"
+        "```erlang formatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "This unformatted version, should match the first formatted version with the same key\n"
+        "```erlang unformatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n",
+    check_markdown(S).
+
+check_markdown(Content) ->
+    Sections = string:split(Content, "```", all),
+    {_, Formatted, Unformatted} = lists:foldl(
+        fun split_code_into_maps/2,
+        {text, maps:new(), maps:new()},
+        Sections
+    ),
+    maps:map(
+        fun (Key, FormattedCode) ->
+            check_fmt(FormattedCode, FormattedCode),
+            case maps:find(Key, Unformatted) of
+                error ->
+                    ignore;
+                {ok, UnformattedCode} ->
+                    check_fmt(UnformattedCode, FormattedCode)
+            end
+        end,
+        Formatted
+    ),
+    % check that there are no unformatted pairs without formatted friends
+    ?assertEqual(
+        0,
+        maps:size(
+            maps:fold(
+                fun (Key, _Value, FriendsLeft) ->
+                    maps:remove(Key, FriendsLeft)
+                end,
+                Unformatted,
+                Formatted
+            )
+        )
+    ).
```

Comment:
I would probably write this as
```suggestion
    ?assertEqual(#{}, maps:without(maps:keys(Formatted), Unformatted)).
```

Reply:
Thank you this is great.
I am learning so much from this review.
Thank you for taking the time.